lib/gc/Transforms/GPU/GpuTilingAndFusion.cpp

```diff
@@ -0,0 +1,481 @@
+//===-- GpuTilingAndFusion.cpp - DESC ---------------------------*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "./GpuUtils.h"
+#include "gc/Dialect/Linalgx/Utils.h"
+
+#include "mlir/Conversion/Passes.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Affine/Utils.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/GPU/Transforms/Passes.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Dialect/Math/IR/Math.h"
+#include "mlir/Dialect/SCF/Transforms/TileUsingInterface.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Transforms/RegionUtils.h"
+
+using namespace mlir;
+using namespace mlir::gc;
+using namespace mlir::scf;
+
+namespace mlir::gc {
+#define GEN_PASS_DECL_GPUTILINGANDFUSION
+#define GEN_PASS_DEF_GPUTILINGANDFUSION
+#include "gc/Transforms/Passes.h.inc"
+} // namespace mlir::gc
+
+namespace {
+
+struct GpuTilingAndFusion final
+    : GpuPass<GpuTilingAndFusion>,
+      gc::impl::GpuTilingAndFusionBase<GpuTilingAndFusion> {
+  friend struct GpuPass;
+  explicit GpuTilingAndFusion()
+      : GpuTilingAndFusion(GpuTilingAndFusionOptions{}) {}
+  explicit GpuTilingAndFusion(const GpuTilingAndFusionOptions &opts)
+      : GpuPass(), GpuTilingAndFusionBase(opts) {}
+
+  void runOnOperation() override {
+    auto fn = getOperation();
+    if (fn.isExternal()) {
+      return;
+    }
+
+    OpRewriter rw(fn);
+    tileAndFuseLinalgOps(rw, fn);
+    tileForallOps(rw, fn);
```

Comment:
The second phase just creates a nested loop, there is nothing to fuse, it just tiles the outer loop.

Reply:
ok, then why do you tile the outer loop manually instead of calling the `tileandFuse` twice?
```
linalg.op

tileAndFuse linalg.op:
scf.forall {
  linalg.op
}

tileAndFuse linalg.op:
scf.forall {
  scf.forall {
    linalg.op
  }
}

```

Reply:
Because we need to tile a single loop just to spread computations over the work-groups. We don't need to tile and fuse all the operations inside the loop. Actually, I tried it in the first approach, it worked for simple cases, like in your example, but for some more complex cases the result was not as it was expected.

Reply:
Do you have a complex case that the result is not expected calling the tileAndFuse twice?

Reply:
What does it mean by **different result**? As far as I know, the `tileConsumerAndFuseProducersUsingSCF` will insert `affine.min, affine.max, affine.apply` to ensure the tiling bound is valid if the dynamic shape exists. The extra code is for safety. There should be no difference in the fusion as the fusion only cares about the slice def-use chain. 

If you want to tile scf.forall only, you could follow the `scf-parallel-loop-tiling` to do it after bufferization which is even easier(https://mlir.llvm.org/docs/Passes/#-scf-parallel-loop-tiling).

Reply:
tileConsumerAndFuseProducersUsingSCF uses the tiles calculation function, provided in the options. There are 2 implementations - one for static and one for dynamic sizes. On the first run of tileConsumerAndFuseProducersUsingSCF the static version will be used, but if dynamic shapes appeared, the dynamic version will be used and the result will be different.